collector/ethtool_linux.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !noethtool
+
+// The hard work of collecting data from the kernel via the ethtool interfaces is done by
+// https://github.com/safchain/ethtool/
+// by Sylvain Afchain. Used under the Apache license.
+
+package collector
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/sysfs"
+	"github.com/safchain/ethtool"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type ethtoolCollector struct {
+	fs      sysfs.FS
+	entries map[string]*prometheus.Desc
+	logger  log.Logger
+}
+
+var (
+	ethtoolFixtures = kingpin.Flag("collector.ethtool.fixtures", "test fixtures to use for ethtool collector end-to-end testing").Default("").String()
+)
+
+func init() {
+	registerCollector("ethtool", defaultDisabled, NewEthtoolCollector)
+}
+
+// NewEthtoolCollector returns a new Collector exposing ethtool stats.
+func NewEthtoolCollector(logger log.Logger) (Collector, error) {
+	fs, err := sysfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %w", err)
+	}
+
+	// Pre-populate some common ethtool metrics.
+	return &ethtoolCollector{
+		fs: fs,
+		entries: map[string]*prometheus.Desc{
+			"rx_bytes": prometheus.NewDesc(
+				"node_ethtool_rx_bytes_total",
```

Comment:
We spell out rx in other exporters, so should do it here too

Reply:
I'm now replacing any "rx" and "tx" in the kernel ethtool names with "received" and "transmitted".